src/crystal/system/win32/file_descriptor.cr

```diff
@@ -10,8 +9,19 @@ module Crystal::System::FileDescriptor
   @volatile_fd : Atomic(LibC::Int)
   @system_blocking = true
 
+  # Residual bytes for character.
+  @console_bytes = Slice(UInt8).empty
+  @console_bytes_buffer : Slice(UInt8)?
+
+  # UTF-16 code units from/to console.
+  @console_units = Slice(UInt16).empty
+  @console_units_buffer : Slice(UInt16)?
```

Comment:
All `IO::FileDescriptor`s and `File`s would become 72 bytes larger (on 64-bit systems) because of this, even ones that aren't console handles. I think these and the other additions should be restricted to the standard streams for the moment

Reply:
I believe it's possible to rewrite the code so that `console_units` is not needed. We may not be able to get rid of a small buffer for residual bytes, but a single instance variable should suffice for that.

Reply:
Yes, I didn't notice that this implementation increases the size of all file descriptors. I think it can be improved like this:

Define a `Console` class:

```
# :nodoc:
class Console
  # Considering that a process has only one console, we can put the status of residual UTF-8 bytes and residual UTF-16 code units into class variables.
  # But I'm not sure if manipulating these class variables requires locking.

  def initialize(
    @handle : LibC::HANDLE,
    @read : (LibC::HANDLE, Void*, LibC::DWORD, LibC::DWORD*, Void*) -> LibC::BOOL = ->LibC.ReadConsoleW,
    @write : (LibC::HANDLE, Void*, LibC::DWORD, LibC::DWORD*, Void*) -> LibC::BOOL = ->LibC.WriteConsoleW
  )
  end

  def read(slice : Bytes) : Int32
    raise "TODO"
  end

  def write(slice : Bytes) : Int32
    raise "TODO"
  end
end
```

Then we use the `Console` like this:
When it is determined that want to read or write from the console, create a `Console` object and then read and write:
```
console = Console.new(handle)
console.read/write
```
So we can test as well: just mock `@read` and `@write` for a `Console` object.

Do you agree to this refactoring?

Reply:
> I believe it's possible to rewrite the code so that `console_units` is not needed. We may not be able to get rid of a small buffer for residual bytes (but maybe we can...?). So I'm not sure if that calls for a specific type for console inputs. But I think this should probably be a dedicated discussion. However, let's figure out the implementation first, because there is potential for improvements.

I seem to see what you mean, do you mean that the UTF-16 code units to read can be calculated by the following formula:
```
# NOTE: Counting the remnants of the last read, at least two UTF-16 code units are required.
units_to_read = slice.size // 3 
```
This ensures that the `slice` doesn't overflow when filling it, so there's no need for `console_units_buffer`, but we need a `StaticArray` or a `Tuple` to store the residual UTF-8 bytes, up to a maximum of 3, and a `UInt16?` to store the residual UTF-16 code unit, right? @straight-shoota

Reply:
I don't see any purpose of the `Console` class.
I'd appreciate if we would focus the architectural discussion in https://github.com/crystal-lang/crystal/issues/13731, I left some comments there on these topics and I believe neither buffer is actually necessary.